From b57befbb23bd4ed6c3c16bb2548eed182c3e91e7 Mon Sep 17 00:00:00 2001
From: Jack Yu <jackcwyu@google.com>
Date: Wed, 30 Jan 2019 15:10:07 +0800
Subject: [PATCH 5/5] Prevent OOB error in rw_i93_sm_update_ndef()

Bug: 122320256
Test: NFC tag reading
Change-Id: Iee56827ff3b65718a61db6cbaca19b4c5abbc223
Merged-In: I053e9f01dc921bab55b5781c83f048d2638d5b87
(cherry picked from commit 8617cbb1a8c2f08de97b6d4c48c053781c67a926)
---
 src/nfc/tags/rw_i93.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/src/nfc/tags/rw_i93.c b/src/nfc/tags/rw_i93.c
index 3b4e253..137fc70 100644
--- a/src/nfc/tags/rw_i93.c
+++ b/src/nfc/tags/rw_i93.c
@@ -2115,6 +2115,13 @@ void rw_i93_sm_update_ndef (BT_HDR *p_resp)
     RW_TRACE_DEBUG1 ("rw_i93_sm_update_ndef () sub_state:0x%x", p_i93->sub_state);
 #endif
 
+    if (length == 0 || p_i93->block_size > I93_MAX_BLOCK_LENGH)
+    {
+        android_errorWriteLog(0x534e4554, "122320256");
+        rw_i93_handle_error(NFC_STATUS_FAILED);
+        return;
+    }
+
     STREAM_TO_UINT8 (flags, p);
     length--;
 
@@ -2144,6 +2151,13 @@ void rw_i93_sm_update_ndef (BT_HDR *p_resp)
         /* get offset of length field */
         length_offset = (p_i93->ndef_tlv_start_offset + 1) % p_i93->block_size;
 
+        if (length < length_offset)
+        {
+            android_errorWriteLog(0x534e4554, "122320256");
+            rw_i93_handle_error(NFC_STATUS_FAILED);
+            return;
+        }
+
         /* set length to zero */
         *(p + length_offset) = 0x00;
 
@@ -2162,6 +2176,12 @@ void rw_i93_sm_update_ndef (BT_HDR *p_resp)
             /* write the first part of NDEF in the same block */
             for ( ; xx < p_i93->block_size; xx++)
             {
+                if (xx > length || p_i93->rw_length > p_i93->ndef_length)
+                {
+                    android_errorWriteLog(0x534e4554, "122320256");
+                    rw_i93_handle_error(NFC_STATUS_FAILED);
+                    return;
+                }
                 if (p_i93->rw_length < p_i93->ndef_length)
                 {
                     *(p + xx) = *(p_i93->p_update_data + p_i93->rw_length++);
@@ -2337,6 +2357,12 @@ void rw_i93_sm_update_ndef (BT_HDR *p_resp)
                 /* update length field within the read block */
                 for (xx = length_offset; xx < p_i93->block_size; xx++)
                 {
+                    if (xx > length)
+                    {
+                        android_errorWriteLog(0x534e4554, "122320256");
+                        rw_i93_handle_error(NFC_STATUS_FAILED);
+                        return;
+                    }
                     if (p_i93->rw_length == 3)
                         *(p + xx) = 0xFF;
                     else if (p_i93->rw_length == 2)
-- 
2.7.4

